ArdalisSpecification/src/Ardalis.Specification/IRepositoryBaseOfT.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Ardalis.Specification
+{
+    /// <summary>
+    /// <para>
+    /// A <see cref="IRepositoryBase{T}" /> can be used to query and save instances of <typeparamref name="T" />.
+    /// An <see cref="ISpecification{T}"/> (or derived) is used to encapsulate the LINQ queries against the database.
+    /// </para>
+    /// </summary>
+    /// <typeparam name="T">The type of entity being operated on by this repository.</typeparam>
+    public interface IRepositoryBase<T> where T : class
+    {
+        /// <summary>
+        /// Adds an entity in the database.
+        /// </summary>
+        /// <param name="entity">The entity to add.</param>
+        /// <returns>
+        /// A task that represents the asynchronous operation.
+        /// The task result contains the <typeparamref name="T" />.
+        /// </returns>
+        Task<T> AddAsync(T entity);
+        /// <summary>
+        /// Updates an entity in the database
+        /// </summary>
+        /// <param name="entity">The entity to update.</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        Task UpdateAsync(T entity);
+        /// <summary>
+        /// Removes an entity in the database
+        /// </summary>
+        /// <param name="entity">The entity to delete.</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        Task DeleteAsync(T entity);
+        /// <summary>
+        /// Removes the given entities in the database
+        /// </summary>
+        /// <param name="entities">The entities to remove.</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        Task DeleteRangeAsync(IEnumerable<T> entities);
+        /// <summary>
+        /// Persists changes to the database.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        Task SaveChangesAsync();
+
+        /// <summary>
+        /// Finds an entity with the given primary key value.
+        /// </summary>
+        /// <param name="id">The value of the primary key for the entity to be found.</param>
+        /// <returns>
+        /// A task that represents the asynchronous operation.
+        /// The task result contains the <typeparamref name="T" />, or <see langword="null"/>.
+        /// </returns>
+        Task<T?> GetByIdAsync(int id);
```

Comment:
If we're using `TId` should we still have an `int` overload?
If we're going to have an `int` overload should we have a `Guid` and/or `string` overload too?

Reply:
I'm not sure even if we need the int overload. the generic version will pick up the type, and actually, it will be inferred, not that you will need to specify `<int>` part. So, probably we don't need any of the specific overloads.

Reply:
Shorter it better, then.

Reply:
Should I remove it completely?